File: dwcj-engine/src/main/java/org/dwcj/component/html/elements/UnorderedList.java
package org.dwcj.component.html.elements;

import org.dwcj.component.Component;
import org.dwcj.component.element.annotation.NodeName;
import org.dwcj.component.html.HtmlComponentContainer;

/**
 * Component representing a {@code ul} element.
 *
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul">HTML ul Tag</a>
 *
 * @author Hyyan Abo Fakher
 * @since 23.06
 */
// We're purposefully ignoring the deep inheritance warning here because we've designed our class
// hierarchy to meet the unique requirements of our UI framework. This design closely aligns with
// our framework's specific goals and emphasizes the need for caution when considering any changes.
//
// Any changes to the inheritance structure should be thoughtfully evaluated in the context of our
// framework's needs. The current structure is essential for meeting those needs.
@SuppressWarnings("squid:S110")
@NodeName("ul")
public class UnorderedList extends HtmlComponentContainer<UnorderedList> {

  /**
   * Creates a new empty UnorderedList.
   */
  public UnorderedList() {
    super();
  }

  /**
   * Creates a new UnorderedList with the given text.
   *
   * @param text the text
   */
  public UnorderedList(String text) {
    super();
    setText(text);
  }

  /**
   * Creates a new UnorderedList with the given child components.
   *
   * @param components the child components
   */
  public UnorderedList(Component... components) {
    super(components);
  }
}

File: webforj-kotlin/src/main/kotlin/com/webforj/kotlin/dsl/component/html/elements/UnorderedList.kt
package com.webforj.kotlin.dsl.component.html.elements

import com.webforj.component.html.elements.ListEntry
import com.webforj.component.html.elements.UnorderedList
import com.webforj.concern.HasComponents
import com.webforj.kotlin.dsl.WebforjDsl
import com.webforj.kotlin.dsl.init

/**
 * Creates an `UnorderedList` with an optional [text].
 * ```
 * ... {
 *   unorderedList() // Empty ul element
 *   unorderedList("text") // ul element with text
 * ```
 *
 * @param text The text to add to the `UnorderedList`.
 * @param block The initialization steps for the `UnorderedList`.
 * @return The configured `UnorderedList`.
 * @see UnorderedList
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ul">HTML ul Tag</a>
 */
fun @WebforjDsl HasComponents.unorderedList(text: String? = null, block: @WebforjDsl UnorderedList.() -> Unit = {}): UnorderedList {
    val ul = text?.let { UnorderedList(text) } ?: UnorderedList()
    return init(ul, block)
}

/**
 * Creates an `ListEntry` for an [UnorderedList] with an optional [text].
 * ```
 * unorderedList {
 *   listEntry() // Empty li element
 *   listEntry("text") // li element with text
 * }
 * ```
 *
 * @param text The text to add to the `ListEntry`.
 * @param block The initialization steps for the `ListEntry`.
 * @return The configured `ListEntry` for the [UnorderedList].
 * @see ListEntry
 * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTML/Element/li">HTML li Tag</a>
 */
fun @WebforjDsl UnorderedList.listEntry(text: String? = null, block: @WebforjDsl ListEntry.() -> Unit = {}): ListEntry {
    val li = text?.let { ListEntry(text) } ?: ListEntry()
    return init(li, block)
}
